src/NHibernate/Impl/AbstractQueryImpl.cs

```diff
@@ -95,11 +95,15 @@ protected internal virtual void VerifyParameters(bool reserveFirstParameter)
 				throw new QueryException("Not all named parameters have been set: " + CollectionPrinter.ToString(missingParams), QueryString);
 			}
 
-			int positionalValueSpan = 0;
-			for (int i = 0; i < values.Count; i++)
+			var positionalValueSpan = 0;
+			// Values and Types may be overriden to yield refined parameters, check them
+			// instead of the fields.
+			var values = Values;
+			var types = Types;
```

Comment:
Trying to use here `ValueArray()` and `TypeArrays()` instead wrecks `CollectionFilterImpl` and `ExpressionFilterImp`, because they override these methods and cause them to yield values failing the validation. So I had to use the `Values` and `Types` properties instead.